man/pycam.1
.TH PyCAM 1 "April 02010" "PyCAM" "PyCAM manual"
.SH NAME
pycam \- start the graphical user interface of the \fBPyCAM\fR Toolpath Generator
for 3-Axis CNC machining
.SH SYNOPSIS
.B pycam
[\fIoptions\fR] [\fISTL_input\fR [\fIGCode_output\fR]]
.SH DESCRIPTION
.PP
\fBPyCAM\fR is a toolpath generator for 3-axis CNC machining. 
.PP
You can configure different cutter shapes and differen path strategies for preparing
multiple passes.
.PP
The output of \fBPyCAM\fR is G-Code suitable for EMC2 or other machine controllers.
.SH OPTIONS
The following options control the behaviour of \fBPyCAM\fR:
.TP
\fB\-h\fR, \fB\-\-help\fR
show the available options and exit
.TP
\fB\-\-gui\fR
don't create the outputfile on the fly - just preset the output filename and show the GUI
.TP
\fB\-\-gtk\fR
use the new GTK interface (default)
.TP
\fB\-\-tk\fR
use the (old) Tk interface
.TP
\fB\-c\fR, \fB\-\-config\fR=\fBCONFIG_FILE\fR
load a task settings file (requires the GTK interface)
.TP
\fB\-\-task\fR=\fBTASK\fR
choose a specific task from task settings by number
(requires the GTK interface)
.TP
\fB\-\-disable-ode\fR
disable experimental ODE collision detection
.SH ENVIRONMENT VARIABLES
.IP PYCAM_DATA_DIR
Override the default location of the GTK interface definition files.
.IP PYTHONPATH
You may want to define this variable in case that you installed the
\fBPyCAM\fR python package in a non-default location.
.SH AUTHOR
Written by Lars Kruse
.SH REPORTING BUGS
See http://sourceforge.net/tracker/?group_id=237831&atid=1104176
.SH COPYRIGHT
Copyright \(co 02010 Lars Kruse
.br
This is free software. You may redistribute copies of it under the terms of the
GNU General Public License <http://www.gnu.org/licenses/gpl.html>. There is NO
WARRANTY, to the extent permitted by law.


Changelog
Version 0.3.1 - UNRELEASED
 * added automatic support grid positioning for contour models
 * allow to reverse the direction of a 2D contour model

Version 0.3.0 - 2010-08-16
 * added support for importing contour paths from SVG files (requires Inkscape and pstoedit)
 * added basic support for importing simple DXF contour files
 * added support for engravings (along the lines of a contour model)
 * added a feature-complete commandline interface
 * integrated "help" links pointing to the wiki
 * improved OpenGL lighting (contributed by imyrek)
 * improved flexibility of the support grid layout:
  * allow non-square profiles for the support grid
  * added ability to create support grids with different x/y grid distance
  * added x/y offsets to support grid options
 * added 3D view movement and rotation with a keyboard in addition to the mouse
 * added optional support for the Psyco just-in-time compiler
 * information about the toolpath settings are stored as comments in GCode files
 * added a log window
 * switched default tool size from radius to diameter
 * improved progress bar responsiveness
 * fixed first movement in GCode: go up to safety height before moving to the side
 * fixed model orientation of STL files created by Art of Illusion
 * fixed performance issue when using a support grid
 * fixed empty toolpath after transforming the model
 * fixed various minor bugs of the toolpath generator
 * fixed rounded corners (see GCode G61)

Version 0.2.5 - 2010-06-10
 * added support bridges for holding the object during cutting
 * calculate the estimated machine time for each toolpath
 * changing the unit (mm/inch) now opens a dialog for scaling the model,
   processing dimensions or tool dimensions accordingly
 * remove unnecessary moves to safety height
 * changed name of configuration setting "overlap" to "overlap_percent"
   (you may need to change this name in your custom config files)

Version 0.2.4 - 2010-04-12
 * added a simple simulation mode for visualizing the material penetration of a toolpath
 * join tangential moves (removes the inner points in a colinear set of adjacent path points)
 * fixed careless import of Tkinter
 * added missing INSTALL.TXT to source package
 * fixed typo that breaks PyCAM for Python 2.6

Version 0.2.3 - 2010-04-05
 * GUI change: tool and process settings can be combined into tasks
 * store configured settings to a file in the user's home directory
 * added "material allowance" support for non-ODE calculations
 * added export of EMC2 tool definitions
 * added a warning dialog (GUI) for missing dependencies during startup (especially useful for Windows)
 * improved GUI for model scaling
 * allow to configure if the tool should move inside/along/around the boundary limits
 * fixed "overlap" calculation
 * prevent invalid input values (zero tool radius, ...)
 * fixed a bug that could rarely cause eternal loops of the PushCutter
 * reduced memory consumption of toolpaths for python 2.6 or above
 * disabled ODE as the default computation backend (due to this bug:
   http://sourceforge.net/tracker/?func=detail&aid=2973876&group_id=24884&atid=382799)

Version 0.2.2 - 2010-03-17
 * added a graphical installer for Windows (via distutils)
 * fixed broken commandline parameter "--template"
 * added workaround for ODE collision detection, that is broken under specific circumstances
 * fixed hang with PushCutter and ODE on Windows
 * fixed "division by zero" error in non-ODE mode
 * allow to disable ODE via commandline option
 * bugs fixed in cylindrical and toroidal cutter when using dropcutter on horizontal triangles
 * fixed "setup.py" for distutils packaging (contributed by Arthur Magill)

Version 0.2.1 - 2010-03-09
 * fixed code that depended on GTK 2.16 (instead of 2.12)
 * view settings "light", "shadow" and "polygon fill" are now configurable
 * documented version problems with Debian "Lenny" and Ubuntu (before "Karmic")

Version 0.2 - 2010-03-04
 * added an alternative GTK interface with additional features:
  * configurable 3D view settings: model / toolpath / axis / drill progress / frame rate / colors
  * manipulation of the model: rotation, flip, swap, scale, move
  * saving the model to an ascii STL file
  * load/save processing settings from/to a file
  * handling of multiple processing templates (e.g. "Rough", "Semi-finish" and "Finish")
  * configure "material allowance" for a minimum distance between drill and model
  * configure the "safety height" for the machine
  * allow to manage and export (gcode) multiple toolpaths at once
  * based on GtkGlExt (for OpenGL)
  * visualize invalid processing setting combinations
  * configure "overlap" instead of "lines" and "samples"
  * configure "maximum step down" instead of "layers"
  * show a progress bar during calculations
  * basic lighting for the OpenGL view
 * improved performance by (optionally) using the "Open Dynamics Engine" (ODE) for
   collision detection

Version 0.1.11b - 2010-02-25
 * fix minor release mistake (missing "STLExporter")

Version 0.1.11 - 2010-02-19
 * fix detection of binary STL
 * added Simulation mode
 * ignore invalid triangles (caused by high-resolotion models)
 * automatic boundary fits to the model
 * add ubuntu specific note
 * code cleanups

Version 0.1.10 - 2009-07-13
 * support binary STL
 * close contours

Version 0.1.9 - 2009-06-18
 * linux compatibility
 * integrated kdtree
 * fix speedup shortcuts

Version 0.1.8 - 2009-04-14
 * some feature requests
 * fixes by Dan Falck
 * fix bug in torus-point intersection
 * specify model in cfg
 * fixed opengl representation
 * update tests

Version 0.1.7b - 2009-02-10
 * fix glutInit bug

Version 0.1.7 - 2009-01-27
 * another try at fixing PushCutter
 * remove MGED exporter

Version 0.1.6b - 2009-01-15
 * fix save
 * follow api updates
 * fix gcode parameter strings

Version 0.1.6 - 2009-01-14
 * fix verticals intersection
 * add ContourCutter
 * guard against float division
 * add x/y direction
 * add ContourCutter

pycam.py
#!/usr/bin/python

from optparse import OptionParser
import sys

# check if we were started as a separate program
if __name__ == "__main__":
    inputfile = None
    outputfile = None

    parser = OptionParser(usage="usage: %prog [options] [inputfile [outputfile]]")
    parser.add_option("", "--gui", dest="display",
            action="store_true", default=False,
            help="don't create the outputfile on the fly - just preset the output filename and show the GUI")
    parser.add_option("", "--gtk", dest="gtk_gui",
            action="store_true", default=True,
            help="use the new GTK interface (default)")
    parser.add_option("", "--tk", dest="gtk_gui",
            action="store_false",
            help="use the (old) Tk interface")
    parser.add_option("-c", "--config", dest="config_file",
            default=None, action="store", type="string",
            help="load a file with processing settings (requires the GTK interface)")
    parser.add_option("", "--template", dest="config_template",
            default=None, action="store", type="string",
            help="enable a specific processing settings template - e.g. 'Rough', 'Semi-finish' or 'Finish' (requires the GTK interface)")
    (options, args) = parser.parse_args()

    if len(args) > 0:
        inputfile = args[0]
    if len(args) > 1:
        outputfile = args[1]
    if len(args) > 2:
        parser.error("too many arguments given (%d instead of %d)" % (len(args), 2))

    # try the configured interface first and then try to fall back to the alternative, if necessary
    if options.gtk_gui:
        try:
            from pycam.Gui.Project import ProjectGui
            gui_class = ProjectGui
        except ImportError, err_msg:
            print >> sys.stderr, "Failed to load GTK bindings for python. Please install the package 'python-gtk2'."
            print >> sys.stderr, "Details: %s" % str(err_msg)
            print >> sys.stderr, "I will try to use the alternative Tk interface now ..."
            try:
                from pycam.Gui.SimpleGui import SimpleGui
                gui_class = SimpleGui()
            except ImportError:
                gui_class = None
    else:
        try:
            from pycam.Gui.SimpleGui import SimpleGui
            gui_class = SimpleGui
        except ImportError, err_msg:
            print >> sys.stderr, "Failed to load TkInter bindings for python. Please install the package 'python-tk'."
            print >> sys.stderr, "Details: %s" % str(err_msg)
            print >> sys.stderr, "I will try to use the alternative GTK interface now ..."
            try:
                from pycam.Gui.Project import ProjectGui
                gui_class = ProjectGui
            except ImportError:
                gui_class = None
    # exit if no interface is found
    if gui_class is None:
        print >> sys.stderr, "Neither the GTK nor the Tk interface is available. Please install the corresponding packages!"
        sys.exit(1)
    else:
        if outputfile:
            gui = gui_class(no_dialog=True)
        else:
            gui = gui_class()

    if not inputfile:
        from pycam.Importers.TestModel import TestModel
        gui.load_model(TestModel())
    else:
        gui.open(inputfile)
    # load processing config file
    # only available for the GTK interface
    if options.gtk_gui:
        if options.config_file:
            gui.open_processing_settings_file(options.config_file)
        if options.config_template:
            gui.switch_processing_config(section=options.config_template)
    if outputfile and not options.display:
        # an output filename is given and no gui is explicitly requested
        gui.generateToolpath()
        if outputfile:
            gui.save_toolpath(outputfile)
    else:
        # the gui should be shown
        if outputfile:
            gui.setOutputFilename(outputfile)
        gui.mainloop()

